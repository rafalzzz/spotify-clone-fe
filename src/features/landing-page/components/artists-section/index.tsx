'use client';
import { FC } from 'react';

import { TArtistSection } from '@/landing-page/types/types';

import { EArtistKeys } from '@/types/artist';

import { CustomSection, CustomArtistItem } from '@/shared/components';

export const ArtistsSection: FC<TArtistSection> = ({ artists }): JSX.Element => (
  <CustomSection title='Most Popular DJs' redirectionUrl='/most-famous-djs'>
    {
      <ul className='custom-section__items'>
        {artists.map((artist) => (
          <li key={artist[EArtistKeys.ARTIST_ID]}>
            <CustomArtistItem
              artistName={artist[EArtistKeys.ARTIST_NAME]}
              artistId={artist[EArtistKeys.ARTIST_ID]}
            />
          </li>
        ))}
      </ul>
    }
  </CustomSection>
);
