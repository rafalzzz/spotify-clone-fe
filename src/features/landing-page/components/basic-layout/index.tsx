import { ReactNode } from 'react';

import Sidebar from '@/sidebar/components/sidebar';

import FooterContainer from '@/footer/components/footer-container';

import Navbar from '@/navigation/components/navbar';

import { TBasicLayout } from '@/landing-page/types/types';

import './BasicLayout.scss';

export const BasicLayout = ({ children }: TBasicLayout) => (
  <div className='basic-layout'>
    <Sidebar />
    <div className='basic-layout__content'>
      <Navbar />
      {children}
    </div>
    <FooterContainer />
  </div>
);
