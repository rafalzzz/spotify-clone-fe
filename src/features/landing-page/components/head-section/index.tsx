'use client';
import Head from 'next/head';
import { FC, PropsWithChildren, useEffect } from 'react';

import { useMusicPlayerStore } from '@/store/music-player';

import { useCurrentSong } from '@/hooks/use-current-song';

import { EMusicTrackKeys } from '@/types/music-track';

import GLOBAL_SETTINGS from '@/configs/global';

export const HeadSection: FC<PropsWithChildren> = (): JSX.Element => {
  const currentSong = useCurrentSong();
  const isPlaying = useMusicPlayerStore(({ isPlaying }) => isPlaying);

  const trackName = currentSong?.[EMusicTrackKeys.TRACK_NAME];
  const artist = currentSong?.[EMusicTrackKeys.ARTIST_NAME];
  const title =
    isPlaying && trackName && artist ? `${trackName} â€¢ ${artist}` : GLOBAL_SETTINGS.DEFAULT_TITLE;

  useEffect(() => {
    document.title = title;
  }, [title]);

  return (
    <Head>
      <title>{title}</title>
    </Head>
  );
};
