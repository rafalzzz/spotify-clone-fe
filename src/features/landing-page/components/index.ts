export { default as ArtistsSection } from './artists-section';
export { default as LoveAlbumsSection } from './love-albums-section';
export { default as LoveSongsSection } from './love-songs-section';
