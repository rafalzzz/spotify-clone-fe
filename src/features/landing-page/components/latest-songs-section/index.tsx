'use client';
import { useCalculateSectionItemsAmount } from '@/hooks/use-calculate-section-items-amount';

import { MusicTrack } from '@/types/music-track';

import { CustomSection, CustomSectionItem, MusicTrackInformations } from '@/shared/components';

export const LatestSongsSection = ({ songs }: { songs: MusicTrack[] }) => {
  const { elementRef, sectionItemsCount } = useCalculateSectionItemsAmount();

  return (
    <CustomSection title='Love songs' redirectionUrl='/love-songs'>
      {
        <ul className='custom-section__items' ref={elementRef}>
          {songs.map(({ trackId, artworkUrl100, trackName, artistName }, index) =>
            index < sectionItemsCount ? (
              <li key={trackId}>
                <CustomSectionItem
                  imageUrl={artworkUrl100}
                  onClick={() => {
                    // TODO - add play song action
                  }}
                >
                  <MusicTrackInformations trackName={trackName} artistName={artistName} />
                </CustomSectionItem>
              </li>
            ) : null,
          )}
        </ul>
      }
    </CustomSection>
  );
};
