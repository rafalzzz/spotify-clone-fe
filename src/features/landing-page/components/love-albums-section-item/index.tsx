'use client';
import { FC, memo, useCallback } from 'react';

import { TLoveAlbumsSectionItem } from '@/landing-page/types/types';

import { useMusicPlayerStore } from '@/store/music-player';

import { CustomSectionItemPlayButton } from '@/components/custom-section-item-play-button';

import { useGetAlbumSongs } from '@/hooks/use-fetch-album-songs';

import { generateAlbumRedirectionPath } from '@/utils/generate-album-redirection-path';

import { EAlbumKeys } from '@/types/album';

import { AlbumInformation, CustomSectionItem, CustomSectionItemImage } from '@/shared/components';

export const LoveAlbumsSectionItem: FC<TLoveAlbumsSectionItem> = memo(
  ({ album, isPlaying, isActive }) => {
    const playAlbum = useMusicPlayerStore(({ playAlbum }) => playAlbum);
    const togglePlay = useMusicPlayerStore(({ togglePlay }) => togglePlay);

    const albumId = album[EAlbumKeys.COLLECTION_ID];

    const { cachedData, fetchAlbumSongsAction } = useGetAlbumSongs({
      albumId,
      onSuccess: (data) => {
        playAlbum({ albumId: album[EAlbumKeys.COLLECTION_ID], songs: data });
      },
    });

    const handleOnClick = useCallback(() => {
      if (isActive) {
        return togglePlay();
      }

      if (cachedData) {
        return playAlbum({ albumId, songs: cachedData });
      }

      fetchAlbumSongsAction.refetch();
    }, [albumId, cachedData, fetchAlbumSongsAction, isActive, playAlbum, togglePlay]);

    return (
      <li key={album[EAlbumKeys.COLLECTION_ID]} data-testid='section-item'>
        <CustomSectionItem href={generateAlbumRedirectionPath(album[EAlbumKeys.COLLECTION_ID])}>
          <>
            <CustomSectionItemImage imageUrl={album[EAlbumKeys.ARTWORK_URL_60]}>
              <CustomSectionItemPlayButton
                isActive={isActive}
                isPlaying={isPlaying}
                onClick={handleOnClick}
              />
            </CustomSectionItemImage>
            <AlbumInformation
              collectionName={album[EAlbumKeys.COLLECTION_NAME]}
              releaseDate={album[EAlbumKeys.RELEASE_DATE]}
              artistName={album[EAlbumKeys.ARTIST_NAME]}
            />
          </>
        </CustomSectionItem>
      </li>
    );
  },
);
