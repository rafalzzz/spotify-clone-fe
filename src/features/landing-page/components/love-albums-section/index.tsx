'use client';

import { TLoveAlbumsSection } from '@/landing-page/types/types';

import { CustomSection, CustomSectionItem, AlbumInformation } from '@/shared/components';

export const LoveAlbumsSection = ({ albums }: TLoveAlbumsSection) => (
  <CustomSection title='Love albums' redirectionUrl='/love-albums'>
    {
      <ul className='custom-section__items'>
        {albums.map(({ collectionId, artworkUrl60, collectionName, releaseDate, artistName }) => (
          <li key={collectionId}>
            <CustomSectionItem
              collectionName={collectionName}
              imageUrl={artworkUrl60}
              onClick={() => {
                // TODO - add play song action
              }}
            >
              <AlbumInformation
                collectionName={collectionName}
                releaseDate={releaseDate}
                artistName={artistName}
              />
            </CustomSectionItem>
          </li>
        ))}
      </ul>
    }
  </CustomSection>
);
