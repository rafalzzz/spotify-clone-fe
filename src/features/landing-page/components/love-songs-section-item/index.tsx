'use client';
import { FC, memo, useCallback } from 'react';

import { TLoveSongsSectionItem } from '@/landing-page/types/types';

import { useMusicPlayerStore } from '@/store/music-player';

import { CustomSectionItemPlayButton } from '@/components/custom-section-item-play-button';

import { EMusicTrackKeys } from '@/types/music-track';

import {
  CustomSectionItem,
  CustomSectionItemImage,
  MusicTrackInformation,
} from '@/shared/components';

export const LoveSongsSectionItem: FC<TLoveSongsSectionItem> = memo(
  ({ song, isPlaying, isActive }) => {
    const changeSong = useMusicPlayerStore(({ changeSong }) => changeSong);
    const togglePlay = useMusicPlayerStore(({ togglePlay }) => togglePlay);

    const handleOnClick = useCallback(() => {
      if (isActive) {
        return togglePlay();
      }

      const songItem = {
        artistName: song[EMusicTrackKeys.ARTIST_NAME],
        trackName: song[EMusicTrackKeys.TRACK_NAME],
        previewUrl: song[EMusicTrackKeys.PREVIEW_URL],
        artworkUrl60: song[EMusicTrackKeys.ARTWORK_URL_60],
      };

      changeSong({ activeIndex: 0, songs: [songItem] });
    }, [isActive, song, changeSong, togglePlay]);

    return (
      <li key={song[EMusicTrackKeys.TRACK_ID]} data-testid='section-item'>
        <CustomSectionItem collectionName={song[EMusicTrackKeys.COLLECTION_NAME]}>
          <>
            <CustomSectionItemImage imageUrl={song[EMusicTrackKeys.ARTWORK_URL_60]}>
              <CustomSectionItemPlayButton
                isActive={isActive}
                isPlaying={isPlaying}
                onClick={handleOnClick}
              />
            </CustomSectionItemImage>
            <MusicTrackInformation
              trackName={song[EMusicTrackKeys.TRACK_NAME]}
              artistName={song[EMusicTrackKeys.ARTIST_NAME]}
            />
          </>
        </CustomSectionItem>
      </li>
    );
  },
);
