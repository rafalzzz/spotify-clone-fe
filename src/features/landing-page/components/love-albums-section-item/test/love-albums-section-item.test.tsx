import { render, fireEvent } from '@testing-library/react';
import React from 'react';
import '@testing-library/jest-dom';

import { useMusicPlayerStore } from '@/store/music-player';

import { EAlbumKeys } from '@/types/album';
import { TSongItem } from '@/types/components';

import { mockSongItem } from '@/consts/mocks';

import { LoveAlbumsSectionItem } from '..';

let mockCachedData: TSongItem[] | null = null;
let mockRefetch = jest.fn();

jest.mock('@/hooks/use-fetch-album-songs', () => ({
  useGetAlbumSongs: () => ({
    cachedData: mockCachedData,
    fetchAlbumSongsAction: { refetch: mockRefetch },
  }),
}));

const mockAlbum = {
  [EAlbumKeys.COLLECTION_ID]: 1,
  [EAlbumKeys.COLLECTION_NAME]: 'AlbumTest',
  [EAlbumKeys.ARTWORK_URL_60]: 'https://www.test-url.com',
  [EAlbumKeys.ARTIST_NAME]: 'ArtistTest',
  [EAlbumKeys.RELEASE_DATE]: '2020-01-01',
};

const PLAY_BUTTON_TEST_ID = 'custom-section-item-play-button';

const renderLoveAlbumsSectionItem = (additionalProps = {}) => {
  const props = {
    album: mockAlbum,
    isPlaying: false,
    isActive: false,
    ...additionalProps,
  };
  return render(<LoveAlbumsSectionItem {...props} />);
};

describe('LoveAlbumsSectionItem', () => {
  it('renders without error', () => {
    const { container } = renderLoveAlbumsSectionItem();
    expect(container).toMatchSnapshot();
  });

  it('calls togglePlay when isActive and the button is clicked', () => {
    const { getByTestId } = renderLoveAlbumsSectionItem({ isActive: true });
    const playButton = getByTestId(PLAY_BUTTON_TEST_ID);
    fireEvent.click(playButton);

    const { togglePlay } = useMusicPlayerStore.getState();
    expect(togglePlay).toHaveBeenCalled();
  });

  it('plays album with cached data when available and the album is not active', () => {
    const { playAlbum } = useMusicPlayerStore.getState();
    mockCachedData = [mockSongItem];

    const { getByTestId } = renderLoveAlbumsSectionItem({ isActive: false });
    const playButton = getByTestId(PLAY_BUTTON_TEST_ID);
    fireEvent.click(playButton);

    expect(playAlbum).toHaveBeenCalledWith({
      albumId: mockAlbum[EAlbumKeys.COLLECTION_ID],
      songs: mockCachedData,
    });
  });

  it('fetches album songs when no cached data is available and the album is not active', () => {
    mockCachedData = null;

    const { getByTestId } = renderLoveAlbumsSectionItem({ isActive: false });
    const playButton = getByTestId(PLAY_BUTTON_TEST_ID);
    fireEvent.click(playButton);

    expect(mockRefetch).toHaveBeenCalled();
  });
});
