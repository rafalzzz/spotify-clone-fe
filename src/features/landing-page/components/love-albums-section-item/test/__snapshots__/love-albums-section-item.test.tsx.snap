// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoveAlbumsSectionItem renders without error 1`] = `
<div>
  <li
    data-testid="section-item"
  >
    <a
      class="custom-section-item__redirection"
      data-testid="custom-section-item-redirection"
      href="/album/albumtest"
    >
      <div
        class="custom-section-item"
      >
        <div
          class="custom-section-item-image"
        >
          <img
            alt="image"
            data-nimg="1"
            decoding="async"
            height="100"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Fwww.test-url.com&w=256&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fwww.test-url.com&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fwww.test-url.com&w=256&q=75 2x"
            style="color: transparent; height: 100%; width: 100%;"
            width="100"
          />
          <button
            class="custom-section-item-play-button "
            data-testid="custom-section-item-play-button"
          >
            <span
              aria-label="play-circle"
              class="anticon anticon-play-circle antd-icon"
              data-testid="custom-section-item-play-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="play-circle"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm144.1 454.9L437.7 677.8a8.02 8.02 0 01-12.7-6.5V353.7a8 8 0 0112.7-6.5L656.1 506a7.9 7.9 0 010 12.9z"
                />
              </svg>
            </span>
          </button>
        </div>
        <div
          class="album-information"
        >
          <div
            data-mouseenterdelay="0.5"
            data-placement="right"
            data-testid="album-information-collection-name-tooltip"
          >
            <h4
              class="album-information__text"
              data-testid="album-information-collection-name"
            >
              AlbumTest
            </h4>
            <span>
              AlbumTest
            </span>
          </div>
          <div
            class="album-information__wrapper"
          >
            <span
              class="album-information__release-date"
              data-testid="album-information-release-date"
            >
              2020
            </span>
            <span
              class="album-information__separator"
            >
              â€¢
            </span>
            <button
              class="album-information__text album-information__artist"
              data-testid="album-information-artist-name"
            >
              <div
                data-mouseenterdelay="0.5"
                data-placement="right"
                data-testid="album-information-artist-name-tooltip"
              >
                <span>
                  ArtistTest
                </span>
                <span>
                  ArtistTest
                </span>
              </div>
            </button>
          </div>
        </div>
      </div>
    </a>
  </li>
</div>
`;
